exercises/circular-buffer/canonical-data.json

```diff
@@ -0,0 +1,296 @@
+{
+  "#": [
+    "In general, these circular buffers are expected to be stateful,",
+    "and each language will operate on them differently.",
+    "Tests tend to perform a series of operations, some of which expect a certain result.",
+    "As such, this common test suite can only say in abstract terms what should be done.",
+    "",
+    "Tests will contain a number of operations. The operation will be specified in the `operation` key.",
+    "Based on the operation, other keys may be present.",
+    "read: Reading from the buffer should produce the item located at `expected`.",
+    "write: Writing the item located at `item` should succeed.",
+    "readShouldFail: Reading from the buffer should fail.",
+    "writeShouldFail: Writing the item located at `item` should fail.",
+    "overwrite: Write the item located at `item` into the buffer, removing the oldest item if necessary.",
+    "clear: Clear the buffer.",
+    "",
+    "Finally, note that all values are integers.",
+    "If your language contains generics, you may consider allowing buffers to contain other types.",
+    "Tests for that are not included here.",
+    ""
+  ],
+  "cases": [
+    {
+      "description": "reading empty buffer should fail",
+      "capacity": 1,
+      "operations": [
+        {
+          "operation": "readShouldFail"
```

Comment:
Even though tracks are free to implement `failure_detected` how they like, I am **guessing** that `failure_detected` as a separate query will *tend* to suggest that it be a separate function. After all, someone porting this exercise to their track might simply think "one function for each operation, right?!", because what reason do they have to think that this `failure_detected` `operation` is any different than the others (`read, write, overwrite, clear`)*?

This hypothetical API with a separate `failure_detected` operation is one I do not wish to encourage porters to implement, especially since students will subsequently see it, and be encouraged to replicate it. I tend to think it's better to make failures clear right when they happen. After all, both queries and commands have the potential of failing, so it seems cumbersome to have to check failure in a separate query. ~~A separate `failure_detected` query also seems difficult to make work in potentially concurrent contexts~~, but luckily that is not a concern for *this* particular exercise!

I have a difficult time fitting in a circular buffer with command-query separation, even without `failure_detected`. I see from https://en.wikipedia.org/wiki/Command%E2%80%93query_separation that command-query separation has the formal definition of "methods should return a value only if they are referentially transparent and hence possess no side effects."

The `read` operation **has a side-effect**, since it **removes the just-read item** from the buffer!
If we wish to follow command-query separation, `read` is not allowed to both return the oldest value and also remove it. 
Instead, we would be forced to separate the `read` operation into `peek` and `dropOldestItem`.
Well, hey, I might not be opposed to that API change, but I unfortunately don't think I will try to convince others that it is the right thing to do for their tracks.

*: Hoewver, I have just realized after writing this argument against `failure_detected` that the current JSON schema also has two "special' `operations`: `readShouldFail` and `writeShouldFail`. What's to tell a porter that *those* two operations are in any way special? ARGH! So is my proposed JSON schema flawed as well?!